### ISSUE TITLE
subscript out of bounds

### ISSUE THREAD

**Issue by tra6sdc** [alexanderrobitzsch/miceadds#14]
df<-structure(list(Zone = structure(c(1L, 2L, 1L, 1L, 1L, 2L, 2L, 
                   3L, 3L), .Label = c("A", "B", "C"), class = "factor"), Y = c(2L, 
                   3L, 5L, 4L, 3L, 5L, 7L, 8L, 10L), Cat1 = structure(c(1L, 1L, 
                   1L, 1L, 1L, 2L, 2L, 2L, 2L), .Label = c("Fact1", "Fact2"), class = "factor"), 
                   Cat2 = structure(c(1L, 1L, 2L, 2L, 2L, 2L, 3L, 3L, 3L), .Label = c("Level1", 
                   "Level2", "Level3"), class = "factor")), class = "data.frame", row.names = c(NA,-9L))
    Fitglm<-glm(Y~Cat1*Cat2,df,family="gaussian")
    summary(Fitglm)
    Fitglm.cluster<-glm.cluster(df,Y~Cat1*Cat2,cluster="Zone",family="gaussian")
    summary(Fitglm.cluster)
    length(Fitglm.cluster$glm_res$coefficients)
    Fitglm.cluster$glm_res$coefficients
    length(Fitglm.cluster$vcov)
    Fitglm.cluster$vcov

This re-produces the error. I think that it is down to the coefficient parameter vector contains the NA's (length 6) but the variance-covaraince doesn't (length 4x4). The summary from Fitglm shows this whilst Fitglm.cluster "crashes" out.

**Comment by alexanderrobitzsch** [comment]
Am 19.10.2018 um 23:09 schrieb tra6sdc:
>      df<-structure(list(Zone = structure(c(1L, 2L, 1L, 1L, 1L, 2L, 2L,
>                     3L, 3L), .Label = c("A", "B", "C"), class = "factor"), Y = c(2L,
>                     3L, 5L, 4L, 3L, 5L, 7L, 8L, 10L), Cat1 = structure(c(1L, 1L,
>                     1L, 1L, 1L, 2L, 2L, 2L, 2L), .Label = c("Fact1", "Fact2"), class = "factor"),
>                     Cat2 = structure(c(1L, 1L, 2L, 2L, 2L, 2L, 3L, 3L, 3L), .Label = c("Level1",
>                     "Level2", "Level3"), class = "factor")), class = "data.frame", row.names = c(NA,-9L))
>      Fitglm<-glm(Y~Cat1*Cat2,df,family="gaussian")
>      summary(Fitglm)
>      Fitglm.cluster<-glm.cluster(df,Y~Cat1*Cat2,cluster="Zone",family="gaussian")
>      summary(Fitglm.cluster)
>      length(Fitglm.cluster$glm_res$coefficients)
>      Fitglm.cluster$glm_res$coefficients
>      length(Fitglm.cluster$vcov)
>      Fitglm.cluster$vcov
> 
> This re-produces the error. I think that it is down to the coefficient parameter vector contains the NA's (length 6) but the variance-covaraince doesn't (length 4x4). The summary from Fitglm shows this whilst Fitglm.cluster "crashes" out.
> 
> 

I guess that it directly pertains `multiwayvcov::cluster.vcov()` because 
`miceadds::glm.cluster()` is just a wrapper around `stats::glm()`and 
`multiwayvcov::cluster.vcov()`. Nevertheless, I can catch the case of 
coefficients with `NA` and produce a warning message. See the source code

https://github.com/alexanderrobitzsch/miceadds/blob/master/R/glm.cluster.R

**Comment by tra6sdc** [comment]
Hi, do I need to post this bug to the multiwayvcov R package? I also note that I get the error even when there are no NA parameter estimates.

    > library(miceadds)
    > df<-structure(list(Zone = structure(c(1L, 2L, 1L, 1L, 1L, 2L, 2L,  +
    3L, 3L), .Label = c("A", "B", "C"), class = "factor"), Y = c(2L,  + 
    3L, 5L, 4L, 3L, 5L, 7L, 8L, 10L), Cat1 = structure(c(1L, 1L,  + 
    1L, 1L, 1L, 2L, 2L, 2L, 2L), .Label = c("Fact1", "Fact2"), class = "factor"),  +  
    Cat2 = structure(c(1L, 1L, 2L, 2L, 2L, 2L, 3L, 3L, 3L), .Label = c("Level1",  + 
    "Level2", "Level3"), class = "factor")), class = "data.frame", row.names = c(NA,-9L))

    > Fitglm<-glm(Y~Cat1+Cat2,df,family="gaussian")
    summary(Fitglm)
    Call: glm(formula = Y ~ Cat1 + Cat2, family = "gaussian", data = df)
    Deviance Residuals:        1        2        3        4        5        6        7        8        9
    -0.5000   0.5000   1.0000   0.0000  -1.0000   0.0000  -1.3333  -0.3333   1.6667
    Coefficients:
             Estimate Std. Error t value Pr(>\|t\|)
    (Intercept)   2.5000     0.8466   2.953   0.0318 * 
    Cat1Fact2     1.0000     1.3824   0.723   0.5019
    Cat2Level2    1.5000     1.0929   1.372   0.2283
    Cat2Level3    4.8333     1.7623   2.743   0.0407 *
    --- Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
    (Dispersion parameter for gaussian family taken to be 1.433333)
     Null deviance: 55.5556  on 8  degrees of freedom
     Residual deviance:  7.1667  on 5  degrees of freedom AIC: 33.491
     Number of Fisher Scoring iteraations 2

    > Fitglm.cluster<-glm.cluster(df,Y~Cat1+Cat2,cluster="Zone",family="gaussian") 
    > summary(Fitglm.cluster)
    Error in `[<-`(`*tmp*`, , "z value", value = csmod[, "Estimate"]/csmod[,  :
      subscript out of bounds 
    > length(Fitglm.cluster$glm_res$coefficients) [1] 4 
    > Fitglm.cluster$glm_res$coefficients 
    (Intercept)   Cat1Fact2  Cat2Level2  Cat2Level3 
    2.500000    1.000000    1.500000    4.833333
    > length(Fitglm.cluster$vcov) [1] 16
    > Fitglm.cluster$vcov
                (Intercept)     Cat1Fact2    Cat2Level2    Cat2Level3
    (Intercept)  3.000000e-01  1.973730e-16 -3.000000e-01 -5.666667e-01
    Cat1Fact2    7.012097e-31  7.012097e-31 -1.402419e-30 -4.715021e-16
    Cat2Level2  -3.000000e-01 -1.973730e-16  3.000000e-01  5.666667e-01
    Cat2Level3  -5.666667e-01 -7.894919e-16  5.666667e-01  1.781481e+00

**Comment by alexanderrobitzsch** [comment]
I replicated your bug. You were indeed right. The `summary.glm.cluster` produced a bug. I fixed the bug in `miceadds 3.0-3`.

**Comment by tra6sdc** [comment]
Thanks Alexander. I emailed the author of multiwayvcov who sent this reply. I expect you are aware of the diagnosis of the issue since you report it as being fixed? Thanks again.

“Stephen,

It looks like whatever's going wrong is happening in summary.glm.cluster(); multiwayvcov is producing a variance-covariance matrix as desired. I narrowed down the error to the assignment on line 35 in glm.cluster.R in miceadds; the division operation is fine if you pull it out and walk through the function line-by-line:

> csmod[, "Estimate"]/csmod[, "Std. Error"]
 (Intercept)    Cat1Fact2   Cat2Level2   Cat2Level3
4.564355e+00 1.194197e+15 2.738613e+00 3.621230e+00

So it looks like the "subscript out of bounds" that's incorrect is "z value", which isn't a valid subscript in the matrix csmod (the generated summary matrix uses t instead of z). Everything works just fine if I change the family of the glm link function to something that uses z scores instead, such as poisson (I mean, the model is trash, but the code works). It looks like this version of summary.glm.cluster has never been tested against gaussian glm models. I don't know anything about miceadds, so I don't know if this is a big deal or not, but it isn't a problem with multiwayvcov.”


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
